### ISSUE TITLE
npm:winston@^2.2.0

### ISSUE THREAD

**Issue by simonpure** [jspm/registry#793]
Adding a specific 

```
require('./transports/console');
require('./transports/file')
```

here at the top

```
https://github.com/winstonjs/winston/blob/master/lib/winston/transports.js
```

seems to fix the issue but not sure why it's required.

```
{ [Error: Module /usr/src/jspm_packages/npm/winston@2.2.0/lib/winston/transports/console not declared as a dependency of file:///usr/src/jspm_packages/npm/winston@2.2.0/lib/winston/transports.js
        Evaluating file:///usr/src/jspm_packages/npm/winston@2.2.0/lib/winston.js
  originalErr: [Error: Module /usr/src/jspm_packages/npm/winston@2.2.0/lib/winston/transports/console not declared as a dependency of file:///usr/src/jspm_packages/npm/winston@2.2.0/lib/winston/transports.js] }
```

**Comment by AdamBurgess** [comment]
also needs `./transports/http`, and `./transports/memory` as dependencies for complete support
continuing on from gitter, @simonpure: this error happens because of a dynamic require which systemjs doesn't support

**Comment by simonpure** [comment]
Thanks @AdamBurgess for the comment and explanation. What's the recommended way to fix this so it plays well in future installs?  I'm assuming this is where overrides come in?

Appreciate any pointers.

**Comment by guybedford** [comment]
The way to fix this would be with a `deps` override, to supplement the static tracing with the necessary info. `deps` overrides for CommonJS is only supported in jspm 0.17, so I'd suggest running jspm 0.17 and using an override like:

```
jspm install npm:winston -o "{meta: { 'lib/winston/transports.js':{ deps:['./http.js','./memory.js','./console.js','./file.js'] } }}"
```

or something along those lines.

**Comment by adamburgess** [comment]
@guybedford another issue, winston was require'ing those files like so:
```js
require(path.join(__dirname, 'transports', 'http.js'))
```
which meant it resolved to a full path e.g. `R:\work\test-winston\jspm_packages\npm\winston@2.2.0\lib\winston\transports\http.js`. Is there a way to get the meta deps to work on a path like this?

---

I've sent a PR to winston which rewrites this transport.js file to use static imports which fixes these 2 problems - https://github.com/winstonjs/winston/pull/884

**Comment by guybedford** [comment]
Meta deps should work on full paths like that as we apply module resolution in fully normalized URL space, which is compatible with that. Perhaps double check the right name is being generated in the registry here? (eg if there's a windows quirk perhaps causing the issue here with `\` instead of `/`?)

Thanks for posting the PR there!

**Comment by adamburgess** [comment]
>there's a windows quirk perhaps causing the issue here with \ instead of /?

`path.join` makes it use \ as delimiters, could that be the problem?

**Comment by guybedford** [comment]
It might be, ideally the main normalization process should convert `C:\asdf\asdf\asdf` into `file:///C:/asdf/asdf/asdf`. Perhaps this is just https://github.com/systemjs/systemjs/issues/1317.

**Comment by adamburgess** [comment]
So, I tried these steps

* replaced the \ with / in the require path: `path.join(__dirname, 'transports', name.toLowerCase()).replace(/\\/g, '/')`

same error:
```
err  (SystemJS) Error: Module R:\Coding\test\jspm_packages\npm\winston@2.2.0\lib\winston\transports\console not declared as a dependency of R:/Coding/test/jspm_packages/npm/winston@2.2.0/lib/winston/transports.js
```

So full paths do not work? Unless this is still just because of a non-URL path, i.e. not `file:///`

* removed __dirname and kept the backslashes: `'.\\transports\\' + name.toLowerCase()`

similar error:
```
err  (SystemJS) Error: Module .\transports\console not declared as a dependency of R:/Coding/test/jspm_packages/npm/winston@2.2.0/lib/winston/transports.js
```

* used slashes: `'./transports/' + name.toLowerCase()`

worked fine.

in all 3 cases my meta override was the same:
```js
"lib/winston/transports.js": {
  "deps": ["./transports/console.js"]
}
```

**Comment by guybedford** [comment]
Ok I think the issue here is slightly deeper.

So basically `R:\...` is being detected as a URI so isn't being normalized into `file:///R:/...` which it should be to work in the URL-based namespace.

I'm not sure what we could do here to make sense of this case though - it unfortunately is one of those things that will work in unix environments but not windows due to the URI detection process.

**Comment by adamburgess** [comment]
>it unfortunately is one of those things that will work in unix environments but not windows due to the URI detection process.

at least I can use bash on windows in the interim :)
```
R:\Coding\test>bash
adam@HELLO:/mnt/r/Coding/test$ node node_modules/jspm/jspm.js run test.js
fullpath: /mnt/r/Coding/test/jspm_packages/npm/winston@2.2.0/lib/winston/transports/console
fullpath: /mnt/r/Coding/test/jspm_packages/npm/winston@2.2.0/lib/winston/transports/console
info: Hello from winston!
adam@HELLO:/mnt/r/Coding/test$ exit
exit

R:\Coding\test>
```

>So basically R:\... is being detected as a URI

not only C:\..., C:/... also doesn't seem to work. Are you saying that you wouldn't plan to make a work around for windows?

**Comment by guybedford** [comment]
The normalization system is URL-based. There's no reliable way to know that a URI protocol isn't intended in the first place when using `:`, unless we start making exceptions for single-letter capital-case protocols...

**Comment by adamburgess** [comment]
Ah, I see your point. There could very well be an `R` or `C` protocol.

**Comment by guybedford** [comment]
Is this sarcasm!?

**Comment by adamburgess** [comment]
well yes and no

it's not sarcasm as in it's completely correct - single letter R could very well be a protocol
but then again, `if OS = windows and protocol.length = 1` is likely to be correct 999 times out of 1000

**Comment by guybedford** [comment]
Yes windows hasn't influenced the normalization system at this level yet.... Having a clear and consistent normalization specification for the loader is important, so exceptions need to be very carefully justified.

I've added a note in https://github.com/systemjs/systemjs/issues/1317, which is basically tracking windows path compatibility. This is also a big pain point in Node dealing with paths between platforms so I agree if we can have a set of rules that work out well things can simplify really nicely.

**Comment by adelespinasse** [comment]
The jspm install command suggested above by @guybedford doesn't have quite the right relative paths in it; this seems to work better:

    jspm install npm:winston -o "{meta: { 'lib/winston/transports.js':{ deps:['./transports/http.js','./transports/memory.js','./transports/console.js','./transports/file.js'] } }}"

(each file is under `./transports/` instead of just `./`)

However, it's still not actually working for me; I get this error now:

```
err  (SystemJS) The super constructor to inherits must not be null or undefined.
  TypeError: The super constructor to inherits must not be null or undefined.
      at Object.exports.inherits (util.js:789:11)
      at Object.<anonymous> (C:/Users/aldel/agora/playdoh/public/jspm_packages/npm/winston@2.2.0/lib/winston/transports/http.js:30:6)
(...etc)
```

which seems to be maybe some kind of failed import. I am using Windows, so I suspect this might be the URL resolution issue, but it wasn't clear to me what symptoms that caused. Is that what I'm seeing?

Is there another way to work around this? Is there a way to tell jspm to use the version in the pull request by @adamburgess, rather than the published version of Winston? I still feel pretty lost trying to configure jspm.

**Comment by adamburgess** [comment]
@adelespinasse afaik I fixed that with something like a 'cjs defer deps' config for the package. I feel like the underlying resolution logic was changed since I last encountered this, so that might work.

**Comment by guybedford** [comment]
Yes there's a `cjsDeferDepsExecute` which can be added to a module meta for these cases to ensure those meta deps aren't executed before the module itself:

```
jspm install npm:winston -o "{meta: { 'lib/winston/transports.js':{ cjsDeferDepsExecute: true, deps:['./transports/http.js','./transports/memory.js','./transports/console.js','./transports/file.js'] } }}"
```

**Comment by adelespinasse** [comment]
tl;dr: Current problem is `exports not defined` in winston/lib/winston/config/cli-config.js; see bottom of post...

@guybedford  That actually didn't work at all. It's back to the original error message about not declaring dependencies on transports/console etc.

Last night I tried to install Winston from @adamburgess's fork (the Winston project seems to be very behind on pull requests). I did something like:

    jspm install winston=github:adamburgess/winston@transports

And then I forced the other package (firebase-queue) that depends on Winston to use that instead of npm:winston (which I ended up doing by directly editing jspm.config.js; I tried to do it with `jspm install -o ...` but apparently I don't understand something about that). It seemed to get further... now the latest error is:

```
err  (SystemJS) exports is not defined
  ReferenceError: exports is not defined
      at C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston/config/cli-config.js:9:17
      at Object.exports.runInThisContext (vm.js:54:17)
      at Object.<anonymous> (C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston/config.js:47:17)
      at C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston/config.js:70:4
      at C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston/config.js:71:3
      at Object.exports.runInThisContext (vm.js:54:17)
  Evaluating C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston/config/cli-config.js
  Evaluating C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston/config.js
  Evaluating C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston/common.js
  Evaluating C:/Users/aldel/agora/playdoh/public/jspm_packages/github/adamburgess/winston@transports/lib/winston.js
  Evaluating C:/Users/aldel/agora/playdoh/public/jspm_packages/npm/firebase-queue@1.5.0/dist/queue.js
  Error loading C:/Users/aldel/agora/playdoh/public/server/server.js
```
Does this even make sense? Is it failing to identify cli-config.js as cjs or something?

**Comment by adelespinasse** [comment]
OH, I needed to specify the registry for winston, because it's [loaded from Github](https://github.com/jspm/registry/wiki/Configuring-Packages-for-jspm#registry). For anyone else who is struggling to make jspm load winston (or any package that depends on it, which in my case was firebase-queue), here are the relevant parts of my project's package.json:

```
{
  "jspm": {
    ...,
    "dependencies": {
      "firebase-queue": "npm:firebase-queue@^1.5.0",
      "winston": "github:adamburgess/winston@transports"
    },
    ...,
    "overrides": {
      ...,
      "npm:firebase-queue@1.5.0": {
        "map": {
          "winston": "github:adamburgess/winston@transports"
        }
      },
      "github:adamburgess/winston@transports": {
        "registry": "npm"
      },
      ...
```

(If you have packages other than firebase-queue that depend on winston, you may have to do something similar with overriding the dependency; I'm not really sure.)

I think none of this will be necessary if the winston project ever accepts @adamburgess's pull request.

**Comment by adelespinasse** [comment]
I think this issue is fixed in version 2.3.0 of winston.
